### PR TITLE
Determinism breaks when the number of workers of a DataLoader

### PR DESCRIPTION
**Pull request by bombs-kim**
### Related PR
Fixed non-determinate preprocessing on DataLoader #4640

### Content
I read the PR above and has always expected that determinism is ensured if I do `manual_seed` before making a data loader. Recently I encountered a situation where the determinism seemed broken. It took me a while to figure out what went wrong, and I learned that the random number generation was conditioned on `worker_id`, hence the number of workers of a loader. I thought it was implicit and hard to expect, so I fixed it.

### An example

```python
import random
from datetime import datetime
import torch
from torch.utils.data import Dataset, DataLoader


class RandomDataset(Dataset):
    def __getitem__(self, item):
        # return torch.LongTensor([random.randint(0, 9999)])
        return torch.LongTensor(1).random_() % 10000

    def __len__(self):
        return 11


def train_load(dset, num_workers):
    # Simulate train loading
    torch.manual_seed(0)  # Fix seed
    loader = DataLoader(dset, num_workers=num_workers)
    print([data.item() for data in loader])


dset = RandomDataset()

for i in range(3):
    train_load(dset, 1)
    
train_load(dset, 2)
train_load(dset, 3)
train_load(dset, 4)

```

```python
# previous (torch 1.0.0)
[8378, 3997, 9259, 9542, 4167, 9875, 2475, 4502, 2503, 2990, 5795]
[8378, 3997, 9259, 9542, 4167, 9875, 2475, 4502, 2503, 2990, 5795]
[8378, 3997, 9259, 9542, 4167, 9875, 2475, 4502, 2503, 2990, 5795]
[8378, 9306, 3997, 4917, 9259, 9434, 9542, 3405, 4167, 2219, 9875]
[8378, 9306, 2250, 3997, 4917, 3659, 9259, 9434, 716, 9542, 3405]
[8378, 9306, 2250, 5787, 3997, 4917, 3659, 2776, 9259, 9434, 716]
```

```python
# this PR
[8378, 9306, 2250, 5787, 5415, 2162, 2645, 4624, 1128, 9602, 7094]
[8378, 9306, 2250, 5787, 5415, 2162, 2645, 4624, 1128, 9602, 7094]
[8378, 9306, 2250, 5787, 5415, 2162, 2645, 4624, 1128, 9602, 7094]
[8378, 9306, 2250, 5787, 5415, 2162, 2645, 4624, 1128, 9602, 7094]
[8378, 9306, 2250, 5787, 5415, 2162, 2645, 4624, 1128, 9602, 7094]
[8378, 9306, 2250, 5787, 5415, 2162, 2645, 4624, 1128, 9602, 7094]
```

@SsnL Please have a look at this PR

### CODE REVIEW

**Review by ssnl — CHANGES_REQUESTED** [review]
I don't think this is correct. You can't assume that indices are integers. You are not doing similar things for single process data loading. Seeding also isn't free, especially if the dataset uses GPU. I feel that the current behavior is reasonable, and what we should do is to make it clear in the docs.

**Comment by bombs-kim** [comment]
@SsnL  Sorry, but why did you think I assumed that indices are integers? I couldn't find which part of my code depends on the condition that indexes be integers

**Comment by bombs-kim** [comment]
@SsnL One another big point of this PR is that, with this implementation we don't have to keep multiple `index_queue`s any more. It's related to @apaszke and @colesbury 's concerns about PR #4640. The problem was that the benchmark testing in that PR was too naive and that PR does seem to deteriorate performance in my case. I'm playing with it now, and found some cases where the performance actually went bad not so trivially.

File: torch/utils/data/dataloader.py [hunk 1]
```diff
@@ -94,7 +94,21 @@ def is_alive(self):
             return not self.manager_dead
 
 
-def _worker_loop(dataset, index_queue, data_queue, done_event, collate_fn, seed, init_fn, worker_id):
+def _fix_dataset(dataset, base_seed):
+    prev_getitem = dataset.__getitem__
+
+    def getitem(self, idx):
+        random.seed(base_seed + idx)
```
Comment (SsnL) [inline torch/utils/data/dataloader.py]: Here you are assuming that index is an integer.

**Comment by SsnL** [comment]
> The problem was that the benchmark testing in that PR was too naive and it does seem to deteriorate performance for me. I'm playing with it now, and found some cases where performance actually went bad not so trivially.

Could you come up with an example along with some benchmark numbers?

> It was surprised by the fact that seeding is not so cheap,

 Seeding is expensive if you are using CUDA.

File: torch/utils/data/dataloader.py (hunk 1 above)
Comment (bombs-kim) [reply torch/utils/data/dataloader.py]: Actually, idx doesn't have to be an integer. See the result of the code below

```python
>> random.seed(1.1);print(random.random())
0.22415427849366876
>> random.seed(1.1);print(random.random())
0.22415427849366876
>>random.seed(1.2);print(random.random())
0.24751702373640583
```

According to the official python docs

> If a(seed) is not None or an int or a long, then hash(a) is used instead

**Comment by bombs-kim** [comment]
@SsnL After I had a conversation with you and learned that torch.manual_seed() is expensive with cuda, I decided that this PR is not worth it any more. It seems that manual seeding for each datapoint is not a good idea to make a deterministically random dataset. Thank you, and I'm happy to close this PR
